crates/types/src/replication/nodeset_selector.rs

```diff
@@ -0,0 +1,1052 @@
+
+        let mut nodes_config = NodesConfiguration::default();
+        nodes_config.upsert_node(generate_node(0, StorageState::Disabled, Role::Admin, ""));
+        nodes_config.upsert_node(generate_node(
+            1,
+            StorageState::Disabled,
+            Role::LogServer,
+            "",
+        ));
+        nodes_config.upsert_node(generate_node(
+            2,
+            StorageState::Provisioning,
+            Role::LogServer,
+            "",
+        ));
+
+        let nodeset = DomainAwareNodeSetSelector::select(
+            &nodes_config,
+            &replication,
+            logserver_candidate_filter,
+            logserver_writeable_node_filter,
+            Default::default(),
+        );
+
+        // provisioning node can be selected, but it's not enough to satisfy total number of
+        // copies.
+        assert_eq!(
+            nodeset,
+            Err(NodeSelectorError::InsufficientNodes {
+                nodeset_len: 1,
+                replication_factor: 2
+            })
+        );
+    }
+
+    /// Replicated loglets should work just fine in single-node clusters, as long as the replication factor is set to 1.
+    #[test]
+    fn select_log_servers_single_node_cluster() {
+        let mut nodes_config = NodesConfiguration::default();
+        nodes_config.upsert_node(generate_node(
+            1,
+            StorageState::ReadWrite,
+            Role::LogServer,
+            "",
+        ));
+
+        let replication =
+            ReplicationProperty::with_scope(LocationScope::Node, 1.try_into().unwrap());
+
+        let preferred_nodes = NodeSet::from_single(1);
+        let options = NodeSetSelectorOptions::new(5).with_preferred_nodes(&preferred_nodes);
+        let selection = DomainAwareNodeSetSelector::select(
+            &nodes_config,
+            &replication,
+            logserver_candidate_filter,
+            logserver_writeable_node_filter,
+            options,
+        );
+
+        assert_eq!(
+            selection.unwrap(),
+            NodeSet::from([1]),
+            "A single-node cluster is possible with replication factor of 1"
+        );
+    }
+
+    /// Replicated loglets should work just fine in single-node clusters, as long as the replication factor is set to 1.
+    #[test]
+    fn select_workers_and_alive_nodes() {
+        let mut nodes_config = NodesConfiguration::default();
+        for i in 1..=100 {
+            nodes_config.upsert_node(generate_node(i, StorageState::Disabled, Role::Worker, ""));
+        }
+
+        // replication = 2, we'll try an have at least 2 writeable nodes in the nodeset. So, we
+        // should expect 5, 8 both to be here.
+        let replication =
+            ReplicationProperty::with_scope(LocationScope::Node, 2.try_into().unwrap());
+        // simulates that we consider 5, 8 are writeable nodes via cluster_state, 5,6 are alive, we prefer to see them in nodeset
```

Comment:
Isn't only 1 in the preferred node set?

Reply:
it's true that 1 is preferred, but 5, 8 are "writeable" so they should get higher rank vs. the rest. So 5,8 are preferred but via cluster_state's choosing them in is_writeable filter.